cpp/include/raft/neighbors/detail/nn_descent.cuh

```diff
@@ -0,0 +1,1447 @@
+}
+
+template <typename Data_t, typename Index_t>
+void GNND<Data_t, Index_t>::local_join(cudaStream_t stream)
+{
+  thrust::fill(thrust::device.on(stream),
+               dists_buffer_.data_handle(),
+               dists_buffer_.data_handle() + dists_buffer_.size(),
+               std::numeric_limits<float>::max());
+  local_join_kernel<<<nrow_, BLOCK_SIZE, 0, stream>>>(
+    thrust::raw_pointer_cast(graph_.h_graph_new.data()),
+    thrust::raw_pointer_cast(h_rev_graph_new_.data()),
+    d_list_sizes_new_.data_handle(),
+    thrust::raw_pointer_cast(h_graph_old_.data()),
+    thrust::raw_pointer_cast(h_rev_graph_old_.data()),
+    d_list_sizes_old_.data_handle(),
+    NUM_SAMPLES,
+    d_data_.data_handle(),
+    ndim_,
+    graph_buffer_.data_handle(),
+    dists_buffer_.data_handle(),
+    DEGREE_ON_DEVICE,
+    d_locks_.data_handle(),
+    l2_norms_.data_handle());
+}
+
+template <typename Data_t, typename Index_t>
+void GNND<Data_t, Index_t>::build(Data_t* data, const Index_t nrow, Index_t* output_graph)
+{
+  using input_t = typename std::remove_const<Data_t>::type;
+
+  cudaStream_t stream = raft::resource::get_cuda_stream(res);
+  nrow_               = nrow;
+  graph_.h_graph      = (InternalID_t<Index_t>*)output_graph;
+
+  cudaPointerAttributes data_ptr_attr;
+  RAFT_CUDA_TRY(cudaPointerGetAttributes(&data_ptr_attr, data));
+  if (data_ptr_attr.type == cudaMemoryTypeUnregistered) {
+    size_t batch_size = 100000;
+    auto input_data   = raft::make_device_matrix<input_t, Index_t, raft::row_major>(
+      res, batch_size, build_config_.dataset_dim);
+    for (size_t step = 0; step < div_up(nrow_, batch_size); step++) {
+      size_t list_offset = step * batch_size;
+      size_t num_lists   = step != div_up(nrow_, batch_size) - 1 ? batch_size : nrow_ - list_offset;
+      raft::copy(input_data.data_handle(),
+                 data + list_offset * build_config_.dataset_dim,
+                 num_lists * build_config_.dataset_dim,
+                 raft::resource::get_cuda_stream(res));
+      preprocess_data_kernel<<<num_lists,
+                               WARP_SIZE,
+                               sizeof(Data_t) * div_up(build_config_.dataset_dim, WARP_SIZE) *
+                                 WARP_SIZE,
+                               stream>>>(input_data.data_handle(),
+                                         d_data_.data_handle(),
+                                         build_config_.dataset_dim,
+                                         l2_norms_.data_handle(),
+                                         list_offset);
+    }
+  } else {
+    preprocess_data_kernel<<<nrow_,
+                             WARP_SIZE,
+                             sizeof(Data_t) * div_up(build_config_.dataset_dim, WARP_SIZE) *
+                               WARP_SIZE,
+                             stream>>>(
+      data, d_data_.data_handle(), build_config_.dataset_dim, l2_norms_.data_handle());
+  }
```

Comment:
We have a `batch_load_iterator` for this. Example: https://github.com/rapidsai/raft/blob/28b789404bedfa8dd82675fc4221f6db927c0422/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L1479-L1492